Improve message alerts

tmaybe opened the pull request:
- [x] Introduces the concept of read and unread messages.
- [x] Indicates clients with unread messages in the index, and sorts them to the top.
- [x] Marks messages read when a client's message index is loaded
- [x] Alerts can now report multiple messages from one or more clients
- [x] The client list updates live as new messages are received
- [x] Messages received while the user's on the sending client's message index are marked read

Closes #12

tmaybe committed:
modularize the twilio sms receiving test code

tmaybe committed:
expand & improve clients controller tests

tmaybe committed:
tests for notifications

Expands twilio helper support functions so they're compatible with
phantomjs/chromedriver.

tmaybe committed:
explicit wait capybara helper function

via @eahanson

tmaybe committed:
expand the authenticated user controller test

tmaybe committed:
replace client controller tests with request tests

Replaced the new controller tests with request tests. The new controller
test that was verifying that a user saw only their own clients is
already covered in a feature test.

tmaybe committed:
styling for clients with unread messages

tmaybe committed:
adding read boolean to messages

tmaybe committed:
test for clients sorting order

We want to sort first by read/unread, second by time of last contact

tmaybe committed:
sent messages are always read

tmaybe committed:
style and sort clients with unread messages

tmaybe committed:
create, not build clients

tmaybe committed:
failing test for marking messages read

tmaybe committed:
messages marked read when user visits index

tmaybe committed:
update factories

tmaybe committed:
test and initial module for alert message building

tmaybe committed:
let read in the message factory default to false

tmaybe committed:
flesh out alert builder tests and functionality

tmaybe committed:
cleanup

tmaybe committed:
make the phone number parser a module

tmaybe committed:
failing tests for new message alerts

Updated twilio_helper to enable the new tests, and updated the tests
that use it with the new process. Also turned the message_alert_builder
into a module.

tmaybe committed:
implement new alert messages, fix tests to pass

tmaybe committed:
failing test, read messages being counted unread

tmaybe committed:
fix read messages included in count

tmaybe committed:
unit test for ignoring read messages

tmaybe committed:
extract clients list into a partial

tmaybe committed:
rename clients partial

tmaybe committed:
don't really need to pass the id

tmaybe committed:
allow custom fake sms text

tmaybe committed:
failing test for live update of client list

tmaybe committed:
cleanup

tmaybe committed:
update the client list in place on new message

tdooner commented on app/controllers/clients_controller.rb:
```diff
@@ -21,4 +26,13 @@ def client_params
+  def sorted_clients
+    # sort clients with unread messages to the top,
+    # no matter when they were last contacted
+    current_user.clients.all.sort{
```
nit: It's idiomatic to use `do` / `end` for multiline blocks, and to also put the block arguments on the first line.

```
.sort do |c1, c2|
  # ...
end
```

tdooner commented on app/assets/javascripts/channels/notifications.js:
```diff
@@ -15,6 +15,13 @@ const Notifications = {
+    // update the client list
+    $.ajax({
+      type: "GET",
+      dataType: "script"
```
Interesting, what are your thoughts about this approach of shipping Javascript versus shipping JSON and then putting the replacement logic here?

tmaybe replied:
I'm trying this out to put [SJR](https://signalvnoise.com/posts/3697-server-generated-javascript-responses) into practice, but it may be a bad test case. I don't really see a great logical advantage of doing it this way in this case.

tdooner commented on app/controllers/clients_controller.rb:
```diff
@@ -21,4 +26,13 @@ def client_params
+    # no matter when they were last contacted
+    current_user.clients.all.sort{
+      |c1, c2| [c2.unread_message_count, c2.contacted_at] <=> [c1.unread_message_count, c1.contacted_at]
+    }
```
Fun ruby tip, you can use `Enumerable#sort_by` to be a bit more efficient because the sort block will be called only once per item instead of this block being called n^2 times.

```ruby
current_user.cilents.all.sort_by { |c| [c.unread_message_count, c.contacted_at] }
```

tdooner commented on spec/lib/message_alert_builder_spec.rb:
```diff
@@ -0,0 +1,66 @@
+      create :message, user: userone, client: client, inbound: true, read: false
+      create :message, user: userone, client: client, inbound: true, read: false
+      expect(described_class.build_alert(user: usertwo)).to eq nil
+    end
```
Nice!

tdooner commented on app/models/client.rb:
```diff
@@ -7,14 +7,32 @@ def full_name
+    if last_received_message
+      return last_received_message.created_at
+    end
+    self.updated_at
```
I'm not sure exactly what interface that the callers of this method are expecting(*), but I would be somewhat surprised if the client has never received a message yet this method returns the `updated_at` timestamp. It seems like making this method return `nil` if the client has never received a message might be less surprising.

\* Am I reading this right that there is nothing calling this method yet?

tmaybe committed:
failing test for messages marked read

Messages should be marked read if they're received when the user's on
the sending client's messages index.

tdooner commented on app/models/client.rb:
```diff
@@ -7,14 +7,32 @@ def full_name
+
+  def unread_message_count
+    # the number of messages received that are unread
+    Message.where(read: false, client: self).count
```
A more reusable approach would be to define a scope for this:

```ruby
# in Messages model:
scope :unread, -> { where(read: false) }

# In Client model
def unread_message_count
  messages.unread.count
end
```

In general, using associations and scopes is more flexible because it allows you to assign semantic meaning to certain queries - so if the implementation of unread messages changes later, it's easy to tell which queries need to change.

It also allows you to compose queries alongside the query DSL, for example:

```
client.messages.unread.sent_by(Devices::PHONE).order(created_at: :desc)
```

Also using associations and scopes allows for more performance optimizations by using activerecord preloader and includes. In general, scopes are awesome and almost all queries should be using a scope or association.

tdooner commented on db/migrate/20170608050028_add_read_to_messages.rb:
```diff
@@ -0,0 +1,11 @@
+    # happens should have read = true
+    Message.all.each do |msg|
+      msg.update_columns read: true
+    end
```
Or just `Message.update_all(read: true)`

tdooner commented on spec/features/user_doesnt_get_alerted_about_read_messages_spec.rb:
```diff
@@ -0,0 +1,38 @@
+      twilio_post_sms(twilio_new_message_params(clientone.phone_number))
+      # there's a flash with the correct content
+      flash_text = "You have 1 unread message from #{clientone.full_name}"
+      expect(page).to have_css '.flash p', text: flash_text
```
```
expect(flash).to eq(flash_text)
```
?

tdooner reviewed (approved):
Looks good, I skimmed most of the tests. Just wanted to give some thoughts on the model association structure, no blockers.

tmaybe committed:
mark messages read if seen by user on arrival

tmaybe committed:
Use relations instead of models to query in Client

This will allow the implementation of these relations to change
without needing to adjust the implementation of the methods that
call them.

tmaybe committed:
test on visible elements instead of db contents

tmaybe committed:
adding a response message to the test for delay

Not good testing practice, but the test needs more time for the ajax
call to successfully mark the client's message as read. So I added and
tested a gratuitous reply to the client to pass time.

tmaybe committed:
up capybara's max wait time

tmaybe committed:
more efficient client sorting

suggested by @tdooner

tmaybe committed:
not using this, so clearing it out

tmaybe committed:
more succinct update in migration

suggested by @tdooner

tmaybe committed:
using the flash helper

suggested by @tdooner

tmaybe committed:
stop using the flash helper

I remembered why I wasn't using it; it wasn't using a capybara function
that auto-waits for the element to appear, which causes intermittent
test failures.
